File: internal/agentrun/event.go
package agentrun

import "time"

// EventModelVersion identifies the closed provider-neutral event contract.
// It is independent of both the project database and AgentRun history formats.
const EventModelVersion uint = 1

// EventKind identifies the observable work product represented by an event.
type EventKind string

const (
	EventLifecycle EventKind = "lifecycle"
	EventTool      EventKind = "tool"
	EventCommand   EventKind = "command"
	EventFile      EventKind = "file"
	EventTest      EventKind = "test"
	EventCommit    EventKind = "commit"
	EventError     EventKind = "error"
	EventSummary   EventKind = "summary"
)

// Valid reports whether the kind belongs to the version 1 contract.
func (k EventKind) Valid() bool {
	switch k {
	case EventLifecycle, EventTool, EventCommand, EventFile,
		EventTest, EventCommit, EventError, EventSummary:
		return true
	default:
		return false
	}
}

// EventPhase describes observable progress without claiming task status.
type EventPhase string

const (
	EventStarted   EventPhase = "started"
	EventProgress  EventPhase = "progress"
	EventWaiting   EventPhase = "waiting"
	EventBlocked   EventPhase = "blocked"
	EventCompleted EventPhase = "completed"
	EventFailed    EventPhase = "failed"
)

// Valid reports whether the phase belongs to the version 1 contract.
func (p EventPhase) Valid() bool {
	switch p {
	case EventStarted, EventProgress, EventWaiting, EventBlocked,
		EventCompleted, EventFailed:
		return true
	default:
		return false
	}
}

// EventOutcome is an optional, explicitly observed operation outcome.
type EventOutcome string

const (
	EventSucceeded    EventOutcome = "succeeded"
	EventUnsuccessful EventOutcome = "failed"
)

// Valid reports whether the outcome belongs to the version 1 contract.
func (o EventOutcome) Valid() bool {
	return o == EventSucceeded || o == EventUnsuccessful
}

// EventNotificationKind is a closed, content-free reason the workspace may
// need the user's attention. ApprovalRequested never represents approval or
// grants any permission.
type EventNotificationKind string

const (
	NotificationApprovalRequested EventNotificationKind = "approvalRequested"
	NotificationQuestion          EventNotificationKind = "question"
	NotificationFailure           EventNotificationKind = "failure"
	NotificationCompletion        EventNotificationKind = "completion"
)

func (k EventNotificationKind) Valid() bool {
	switch k {
	case NotificationApprovalRequested, NotificationQuestion,
		NotificationFailure, NotificationCompletion:
		return true
	default:
		return false
	}
}

// EventObservation is the narrow provider-normalized input contract. It has
// no project, plan, task, terminal, capability, or credential authority. The
// host validates and normalizes these allowlisted fields before creating an
// Event. Summary is explicitly untrusted agent-provided text; later ingestion
// policy bounds and redacts it before any retention.
//
// Subject is a short metadata label such as a tool name, command executable,
// test target, or file operation. It is never a command line, tool input,
// prompt, model response, terminal output, or hidden reasoning.
type EventObservation struct {
	ModelVersion   uint                  `json:"modelVersion"`
	SourceID       string                `json:"sourceId"`
	SourceSequence uint64                `json:"sourceSequence"`
	Kind           EventKind             `json:"kind"`
	Phase          EventPhase            `json:"phase"`
	Outcome        EventOutcome          `json:"outcome,omitempty"`
	Subject        string                `json:"subject,omitempty"`
	Paths          []string              `json:"paths,omitempty"`
	CommitSHA      string                `json:"commitSha,omitempty"`
	ExitCode       *int                  `json:"exitCode,omitempty"`
	ErrorClass     string                `json:"errorClass,omitempty"`
	Summary        string                `json:"summary,omitempty"`
	OccurredAt     time.Time             `json:"occurredAt,omitempty"`
	Notification   EventNotificationKind `json:"notification,omitempty"`
	// recognizedNotification is set only by a provider adapter or trusted
	// history restore. Direct observations cannot self-assert attention.
	recognizedNotification bool
}

// Event is the canonical host-stamped record. RunID, Provider, HostSequence,
// and ObservedAt are assigned from host-owned runtime state; a provider
// observation can never supply or override them.
type Event struct {
	ModelVersion   uint   `json:"modelVersion"`
	ID             string `json:"id"`
	RunID          string `json:"runId"`
	Provider       string `json:"provider"`
	SourceID       string `json:"sourceId"`
	SourceSequence uint64 `json:"sourceSequence"`
	HostSequence   uint64 `json:"hostSequence"`
	// LifecycleRevision is a host-owned epoch. Evidence from a stale lease
	// cannot describe a later heartbeat-revived lifecycle of the same run ID.
	LifecycleRevision uint64                `json:"lifecycleRevision"`
	Kind              EventKind             `json:"kind"`
	Phase             EventPhase            `json:"phase"`
	Outcome           EventOutcome          `json:"outcome,omitempty"`
	Subject           string                `json:"subject,omitempty"`
	Paths             []string              `json:"paths,omitempty"`
	CommitSHA         string                `json:"commitSha,omitempty"`
	ExitCode          *int                  `json:"exitCode,omitempty"`
	ErrorClass        string                `json:"errorClass,omitempty"`
	Summary           string                `json:"summary,omitempty"`
	OccurredAt        time.Time             `json:"occurredAt,omitempty"`
	ObservedAt        time.Time             `json:"observedAt"`
	Correlation       EventCorrelation      `json:"correlation"`
	Notification      EventNotificationKind `json:"notification,omitempty"`
}

File: crates/ptrack-agent/src/event.rs
use serde::{Deserialize, Serialize};

use crate::{EventCorrelation, Timestamp};

pub const EVENT_MODEL_VERSION: u32 = 1;

macro_rules! event_enum {
    ($name:ident { $($variant:ident => $value:literal),+ $(,)? }) => {
        #[derive(Clone, Copy, Debug, Default, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
        pub enum $name {
            #[default]
            #[serde(rename = "")]
            Unset,
            $(#[serde(rename = $value)] $variant),+
        }
        impl $name {
            #[must_use]
            pub const fn is_valid(self) -> bool { !matches!(self, Self::Unset) }
            #[must_use]
            pub const fn as_str(self) -> &'static str {
                match self { Self::Unset => "", $(Self::$variant => $value),+ }
            }
        }
    };
}

event_enum!(EventKind {
    Lifecycle => "lifecycle", Tool => "tool", Command => "command", File => "file",
    Test => "test", Commit => "commit", Error => "error", Summary => "summary",
});
event_enum!(EventPhase {
    Started => "started", Progress => "progress", Waiting => "waiting",
    Blocked => "blocked", Completed => "completed", Failed => "failed",
});
event_enum!(EventOutcome { Succeeded => "succeeded", Failed => "failed" });
event_enum!(EventNotificationKind {
    ApprovalRequested => "approvalRequested", Question => "question",
    Failure => "failure", Completion => "completion",
});

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EventObservation {
    pub model_version: u32,
    pub source_id: String,
    pub source_sequence: u64,
    pub kind: EventKind,
    pub phase: EventPhase,
    #[serde(default, skip_serializing_if = "is_unset_outcome")]
    pub outcome: EventOutcome,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub subject: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub paths: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub commit_sha: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exit_code: Option<i32>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub error_class: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub summary: String,
    #[serde(default, skip_serializing_if = "is_zero_timestamp")]
    pub occurred_at: Timestamp,
    #[serde(default, skip_serializing_if = "is_unset_notification")]
    pub notification: EventNotificationKind,
    #[serde(skip)]
    pub(crate) recognized_notification: bool,
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Event {
    pub model_version: u32,
    pub id: String,
    pub run_id: String,
    pub provider: String,
    pub source_id: String,
    pub source_sequence: u64,
    pub host_sequence: u64,
    pub lifecycle_revision: u64,
    pub kind: EventKind,
    pub phase: EventPhase,
    #[serde(default, skip_serializing_if = "is_unset_outcome")]
    pub outcome: EventOutcome,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub subject: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub paths: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub commit_sha: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exit_code: Option<i32>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub error_class: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub summary: String,
    #[serde(default, skip_serializing_if = "is_zero_timestamp")]
    pub occurred_at: Timestamp,
    pub observed_at: Timestamp,
    pub correlation: EventCorrelation,
    #[serde(default, skip_serializing_if = "is_unset_notification")]
    pub notification: EventNotificationKind,
}

pub(crate) fn observation_from_persisted_event(event: &Event) -> EventObservation {
    EventObservation {
        model_version: event.model_version,
        source_id: event.source_id.clone(),
        source_sequence: event.source_sequence,
        kind: event.kind,
        phase: event.phase,
        outcome: event.outcome,
        subject: event.subject.clone(),
        paths: event.paths.clone(),
        commit_sha: event.commit_sha.clone(),
        exit_code: event.exit_code,
        error_class: event.error_class.clone(),
        summary: event.summary.clone(),
        occurred_at: event.occurred_at,
        notification: event.notification,
        recognized_notification: event.notification.is_valid(),
    }
}

const fn is_unset_outcome(value: &EventOutcome) -> bool {
    matches!(value, EventOutcome::Unset)
}

const fn is_unset_notification(value: &EventNotificationKind) -> bool {
    matches!(value, EventNotificationKind::Unset)
}

const fn is_zero_timestamp(value: &Timestamp) -> bool {
    value.is_zero()
}
